storage/reader.go

```diff
@@ -392,6 +574,72 @@ func (r *Reader) Read(p []byte) (int, error) {
 	return n, err
 }
 
+func (r *Reader) readWithGRPC(p []byte) (int, error) {
+	// No stream to read from, either never initiliazed or Close was called.
+	// TODO: should an error be returned here? "reader has been closed"?
+	if r.stream == nil {
+		return 0, io.EOF
+	}
+
+	// The entire object has been read by this reader, return EOF.
+	if r.size != 0 && r.size == r.seen {
+		return 0, io.EOF
+	}
+
+	n := copy(p, r.leftovers)
+	r.seen += int64(n)
+	r.leftovers = r.leftovers[n:]
+
+	for len(p[n:]) > 0 {
+		msg, err := r.stream.Recv()
```

Comment:
So it looks like our strategy here is to try to fill up `p`. In the Reader interface it says "If some data is available but not len(p) bytes, Read conventionally returns what is available instead of waiting for more." I'm wondering if we should prefer this strategy (which would mean putting off the new Recv call unless leftovers is empty). What do you think?

Reply:
Good idea. I've returned early after exhausting the `leftovers`. It doesn't change much in the way of buffer management, but it does make understanding the code easier, and I like conforming to the documented Reader semantics.

Reply:
Nice. I think in this case we no longer need a loop here either, right? We can just `Recv` once and if that doesn't fill the buffer up, that's fine.

Reply:
You are right! Except we need _some_ loop to Retry the Recv. I've refactored the Recv into `recvWithRetry`.

Reply:
Actually I think this is incorrect-- `reopen` itself does a retry on the `Recv()`. So we shouldn't wrap reopen in a retry as well.

Reply:
Oops you are correct. Fixed!